pkg/plugins/runtime/k8s/controllers/inbound_converter.go

```diff
@@ -158,9 +164,66 @@ func (i *InboundConverter) InboundInterfacesFor(ctx context.Context, zone string
 
 		ifaces = append(ifaces, inboundForServiceless(zone, pod, name, nodeLabels))
 	}
+
 	return ifaces, nil
 }
 
+// Should be used when MeshService mode is Exclusive
+func (i *InboundConverter) InboundInterfacesFor(ctx context.Context, zone string, pod *kube_core.Pod, services []*kube_core.Service) ([]*mesh_proto.Dataplane_Networking_Inbound, error) {
+	nodeLabels, err := i.getNodeLabelsToCopy(ctx, pod.Spec.NodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	var ifaces []*mesh_proto.Dataplane_Networking_Inbound
+	for _, svc := range services {
+		// Services of ExternalName type should not have any selectors.
+		// Kubernetes does not validate this, so in rare cases, a service of
+		// ExternalName type could point to a workload inside the mesh. If this
+		// happens, we would incorrectly generate inbounds including
+		// ExternalName service. We do not currently support ExternalName
+		// services, so we can safely skip them from processing.
+		if svc.Spec.Type != kube_core.ServiceTypeExternalName {
+			ifaces = append(ifaces, inboundForService(zone, pod, svc, nodeLabels)...)
+		}
+	}
+
+	if len(ifaces) == 0 {
+		name, _, err := i.NameExtractor.Name(ctx, pod)
+		if err != nil {
+			return nil, err
+		}
+
+		ifaces = append(ifaces, inboundForServiceless(zone, pod, name, nodeLabels))
+	}
+
+	// Right now we will build multiple inbounds for each service selecting port, but later on
+	// we will only create one listener for last inbound. Ignoring other inbounds from dataplane.
+	// Because of this we can safely deduplicate them here. This needs to change when we get rid of kuma.io/service
+	return deduplicateInboundsByAddressAndPort(ifaces), nil
+}
+
+func deduplicateInboundsByAddressAndPort(ifaces []*mesh_proto.Dataplane_Networking_Inbound) []*mesh_proto.Dataplane_Networking_Inbound {
+	inboundKey := func(iface *mesh_proto.Dataplane_Networking_Inbound) string {
+		return fmt.Sprintf("%s:%d", iface.Address, iface.Port)
+	}
+
+	inboundsPerName := map[string]*mesh_proto.Dataplane_Networking_Inbound{}
+	for _, iface := range ifaces {
+		if inboundsPerName[inboundKey(iface)] == nil {
+			inboundsPerName[inboundKey(iface)] = iface
```

Comment:
append to the result slice here, that way you preserve the original order and you won't need another loop on `inboundsPerName`

Reply:
changed